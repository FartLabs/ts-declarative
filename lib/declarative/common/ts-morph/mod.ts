export * from "./ts-morph.ts";
