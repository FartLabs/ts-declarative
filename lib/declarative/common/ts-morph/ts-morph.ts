import type { Project, SourceFile } from "ts-morph";
import type { Class, Declarative } from "#/lib/declarative/declarative.ts";
import { createDecoratorFactory } from "#/lib/declarative/decorator.ts";

/**
 * TsMorph is a declarative class type from ts-morph.
 */
export interface TsMorph {
  properties: TsMorphProperty[];
}

export interface TsMorphProperty {
  name: string;
  type: string;
  paramIndex?: number;
  // TODO: Add documentation.
  // TODO: Add which class the inherited properties associate with respectively.
}

export interface ValueTsMorph {
  tsMorph?: TsMorph;
}

export function tsMorphDecoratorFactory(
  project: Project,
): (specifier: string | URL) => (target: Class) => Class {
  return createDecoratorFactory({
    initialize: (specifier: URL | string) => {
      const sourceFile = project.getSourceFileOrThrow(specifier.toString());
      return [declarativeTsMorph(sourceFile)];
    },
  });
}

export function declarativeTsMorph<TValue extends ValueTsMorph>(
  sourceFile: SourceFile,
): Declarative<TValue> {
  return (value, name) => {
    return { ...value, tsMorph: getTsMorph(sourceFile, name) } as TValue;
  };
}

export interface ValueType {
  type?: string[];
}

export const type: (
  ...args: Array<string | string[]>
) => (target: Class) => Class = createDecoratorFactory({
  initialize: (...type: Array<string | string[]>) => {
    return [declarativeType(...type)];
  },
});

export function declarativeType<TValue extends ValueType>(
  ...type: Array<string | string[]>
): Declarative<TValue> {
  return <TValue extends ValueType>(value: TValue | undefined): TValue => {
    return { ...value, type: type.flat() } as TValue;
  };
}

function getTsMorph(sourceFile: SourceFile, name: string): TsMorph {
  const classDeclaration = sourceFile.getClass(name);
  const propertyDeclarations = classDeclaration?.getProperties() ?? [];
  const constructorParameterDeclarations =
    classDeclaration?.getConstructors().at(-1)?.getParameters() ?? [];

  return {
    properties: [
      ...propertyDeclarations.map((property): TsMorphProperty => {
        return {
          name: property.getName(),
          type: property.getType().getText(),
        };
      }),
      ...constructorParameterDeclarations.reduce<TsMorphProperty[]>(
        (acc, parameter, i) => {
          if (parameter.getScope() === "public") {
            acc.push({
              name: parameter.getName(),
              type: parameter.getType().getText(),
              paramIndex: i,
            });
          }

          return acc;
        },
        [],
      ),
    ],
  };
}
