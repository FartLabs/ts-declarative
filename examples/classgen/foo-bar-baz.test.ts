import { assertSnapshot } from "@std/testing/snapshot";
import { Project } from "ts-morph";
import { transform, withContextDecorator } from "./classgen.ts";

const fileFooBarBaz = "./examples/classgen/foo-bar-baz.ts";

Deno.test("transform example foo-bar-baz.ts", async (t) => {
  const project = new Project();
  const sourceFile = project.addSourceFileAtPath(fileFooBarBaz);
  transform({ project });
  await assertSnapshot(t, sourceFile.getText());
});

Deno.test("transform example foo-bar-baz.ts with decorators", async (t) => {
  const project = new Project();
  const sourceFile = project.addSourceFileAtPath(fileFooBarBaz);
  transform({
    project,
    map: (structure, sourceDeclarations) => {
      return withContextDecorator(structure, sourceDeclarations);
    },
  });

  await assertSnapshot(t, sourceFile.getText());
});
