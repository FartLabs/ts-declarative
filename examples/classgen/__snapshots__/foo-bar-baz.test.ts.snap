export const snapshot = {};

snapshot[`transform example foo-bar-baz.ts 1`] = `
"class Foo {
    foo: string;
}

class FooBar {
    foo: string;
    bar: number;
}

export class FooBarBaz {
    baz: boolean;
    foo: string;
    bar: number;
}

class Bar {
    bar: number;
}
"
`;

snapshot[`transform example foo-bar-baz.ts with decorators 1`] = `
'@context({ "foo": "Foo/foo" })
class Foo {
    foo: string;
}

@context({ "foo": "Foo/foo", "bar": "Bar/bar" })
class FooBar {
    foo: string;
    bar: number;
}

@context({ "baz": "FooBarBaz/baz", "foo": "Foo/foo", "bar": "Bar/bar" })
export class FooBarBaz {
    baz: boolean;
    foo: string;
    bar: number;
}

@context({ "bar": "Bar/bar" })
class Bar {
    bar: number;
}
'
`;
